spark/src/main/scala/geotrellis/spark/io/FilteringLayerReader.scala

```diff
@@ -1,17 +1,26 @@
 package geotrellis.spark.io
 
 import geotrellis.spark.Boundable
+import geotrellis.spark.io.index.KeyIndex
+import geotrellis.spark.io.json._
+import spray.json.JsonFormat
 
-abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, ReturnType] {
+abstract class FilteringLayerReader[ID, K: Boundable, M, ReturnType] extends LayerReader[ID, K, ReturnType] {
 
-  def read(id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int): ReturnType
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int, format: JsonFormat[I]): ReturnType
+
+  def read[I <: KeyIndex[K]: JsonFormat](id: ID, numPartitions: Int, format: JsonFormat[I]): ReturnType =
```

Comment:
Here's an idea. We have the the `JsonFormat[KeyIndex[K]]` as a constructor parameter. It's never really used because the object applys will pass in the default one (through implicit resolution):

```
object SomethingLayerReader {
  def apply[K,V,M](path: String)(implicit keyIndexFormat: JsonFormat[KeyIndex[K]]) =
    new SomethingLayerReader(path, keyIndexFormat)
}
```

but we have something like a `withKeyIndexFormat` that allows the user to supply a custom format type (if they don't use the implicit resolution to define a newer one that is higher on the implicit scope). So it would look like

```scala
SomethingLayerReader(path).withKeyIndexFormat(customKeyIndexFormatter)
```

That way they could pass in a new formatter that the layer reader could use. If build up some API like Eugene is suggesting to make it extendable, then the user could create a JsonFormat[KeyIndex[K]] by adding new "type" property resolutions to it. Something like

```scala

implicit val customKeyIndexFormat: JsonFormat[KeyIndex[K]] =
  KeyIndexFormat.addType[K, CustomKeyIndexFormat[K]]("custom")
```

where that addType would look for the implicit type-specific resolution and add it to a Keyindex format with the new type tag:

```scala
object KeyIndexFormat {
  def addType[K, I <: KeyIndex[K]: JsonFormat](typeName: String)(implicit generalKeyIndexFormat: JsonFormat[KeyIndex[K]]): JsonFormat[KeyIndex[K]] =
    generalKeyIndexFormat.extent(typeName, implicitly[JsonFormat[I]])
}
```